gui/widgets/qopenglmousetracker.cpp

```diff
@@ -0,0 +1,251 @@
+              rendering::Utils::GLDebugMessageCallback(
+                  msg.source(), msg.type(), msg.id(), msg.severity(),
+                  msg.message().toStdString().c_str());
+            });
+    m_debugLogger->startLogging(QOpenGLDebugLogger::SynchronousLogging);
+  } else {
+    SPDLOG_INFO("QOpenGLDebugLogger was NOT initialized!");
+  }
+
+#endif
+
+  std::string ext =
+      QString::fromLatin1(
+          (const char *)context()->functions()->glGetString(GL_EXTENSIONS))
+          .replace(' ', "\n\t")
+          .toStdString();
+  CheckOpenGLError("glGetString(GL_EXTENSIONS)");
+
+  std::string vendor((char *)context()->functions()->glGetString(GL_VENDOR));
+  CheckOpenGLError("glGetString(GL_VENDOR)");
+  std::string renderer(
+      (char *)context()->functions()->glGetString(GL_RENDERER));
+  CheckOpenGLError("glGetString(GL_RENDERER)");
+  std::string gl_version(
+      (char *)context()->functions()->glGetString(GL_VERSION));
+  CheckOpenGLError("glGetString(GL_VERSION)");
+
+  SPDLOG_INFO("OpenGL: " + vendor + std::string(" ") + renderer +
+              std::string(" ") + gl_version + std::string(" ") +
+              std::string("\n\n\t") + ext + std::string("\n"));
+
+  mainProgram =
+      std::unique_ptr<rendering::ShaderProgram>(new rendering::ShaderProgram(
+          rendering::text_vertex, rendering::text_fragment));
+}
+
+void QOpenGLMouseTracker::renderScene(float lineWidth) {
+  for (color_mesh &obj : meshSet) {
+    obj.second->Render(mainProgram, lineWidth);
+  }
+}
+
+void QOpenGLMouseTracker::paintGL() {
+
+  if (!context()->isValid()) {
+    SPDLOG_ERROR("Widget context is not valid");
+    return;
+  }
+
+  if (QOpenGLContext::currentContext() != context()) {
+    SPDLOG_ERROR("Widget Context is not the active one!");
+    return;
+  }
+
+  camera.UpdateView(mainProgram);
+  camera.UpdateProjection(mainProgram);
+
+  renderScene(lineWidth);
+
+  QOpenGLFramebufferObject fboPicking(size());
+  fboPicking.bind();
+
+  renderScene(lineSelectPrecision);
+
+  offscreenPickingImage = fboPicking.toImage();
+
+  fboPicking.bindDefault();
+}
+
+void QOpenGLMouseTracker::resizeGL(int w, int h) {
+  camera.SetFrustum(camera.getFOV(), w, h, camera.getNear(), camera.getFar());
+  this->update();
+}
+
+void QOpenGLMouseTracker::SetCameraFrustum(GLfloat FOV, GLfloat width,
+                                           GLfloat height, GLfloat nearZ,
+                                           GLfloat farZ) {
+  camera.SetFrustum(FOV, width, height, nearZ, farZ);
+}
+
+void QOpenGLMouseTracker::mousePressEvent(QMouseEvent *event) {
+
+  m_xAtPress = event->position().x();
+  m_yAtPress = event->position().y();
+
+  m_xAtPress = std::clamp(m_xAtPress, 0, offscreenPickingImage.width() - 1);
+  m_yAtPress = std::clamp(m_yAtPress, 0, offscreenPickingImage.height() - 1);
+
+  lastColour = offscreenPickingImage.pixel(m_xAtPress, m_yAtPress);
+  QColor color(lastColour);
+
+  bool objectSelected = false;
+
+#ifdef QT_DEBUG
```

Comment:
no need to wrap logging statements in QT_DEBUG, they can be
- SPDLOG_INFO
- SPDLOG_DEBUG
- SPDLOG_TRACE
as appropriate - all of the above disappear in a build with logging set to WARNING or higher